source/agora/consensus/validation/Block.d

```diff
@@ -143,6 +159,57 @@ public string isInvalidReason (const ref Block block, Height prev_height,
             return fail_reason;
     }
 
+    Point sum_K;
+    Point sum_R;
+    const Scalar challenge = hashFull(block);
+
+    log.trace("Validators expected to sign this block is {}", enrolled_validators);
+    foreach (idx; 0 .. enrolled_validators)
+    {
+        const K = getValidatorAtIndex(block.header.height, idx);
+        log.trace("idx {} PublicKey: {}", idx, PublicKey(K[]));
+        if (K == Point.init)
+        {
+            log.trace("[{}:{}] idx #{}: Validator {} not found for height {}",
+                file, line, idx, PublicKey(K[]), block.header.height);
+            return "Block: Couldn't find a Validator for the given index "
+                ~ to!string(idx) ~ " at height " ~ to!string(block.header.height.value);
+        }
+        if (!block.header.validators[idx])
+        {
+            log.trace("[{}:{}] idx #{}: Validator {} has not yet signed block height {}",
+                file, line, idx, PublicKey(K[]), block.header.height);
+            continue;  // this validator hasn't signed yet
+        }
+
+        const CR = getCommitmentNonce(K);  // commited R
+        log.trace("Commited R for validator {} is {}", PublicKey(K[]), CR);
+        if (CR == Point.init)
+            return "Block: Couldn't find commitment for this validator";
+        Point R = CR + challenge.toPoint();
+        sum_K = sum_K == Point.init ? K : (sum_K + K);
+        sum_R = sum_R == Point.init ? R : (sum_R + R);
+    }
+    if (sum_K == Point.init)
+    {
+        log.error("[{}:{}] Block: Not able to check the multi sig schnorr signature for any of the {} active validators at height {}",
+            file, line, active_enrollments, block.header.height);
+        return "Block: Not enough info to verify schnorr signature at height " ~ to!string(block.header.height.value);
+    }
+    Sig blockSig_Rs = Sig.fromBlob(block.header.signature);
+    if (sum_R != blockSig_Rs.R)
+    {
+        log.error("[{}:{}] Block: Height {} has invalid schnorr signature - sum of R not matching signature R. " ~
+            "enrolled_validators is {} bitmask is {}.",
+            file, line, block.header.height, enrolled_validators, block.header.validators);
+        return "Block: Invalid schnorr signature - sum of R not matching sig";
+    }
+    if (!multiSigVerify(block.header.signature, sum_K, challenge))
+    {
+        log.error("[{}:{}] Block: Invalid schnorr signature for {} active validators",
+            file, line, block.header.validators.length);
+        return "Block: Invalid schnorr signature";
+    }
```

Comment:
Where is the 50% + 1 check? It needs to be included in the validation of the block.

Reply:
Will add as first check.

Reply:
I think you could just count the number of bits set in the bitmask. And then the signature validation will do the rest.

Reply:
Updated